File: it/src/main/scala/org/zalando/nakadi/client/examples/scala/EventCreationExample.scala
package org.zalando.nakadi.client.examples.scala

import org.zalando.nakadi.client.scala.Client
import org.zalando.nakadi.client.scala.ClientFactory
import org.zalando.nakadi.client.scala.model.Event
import org.zalando.nakadi.client.scala.model.EventType
import org.zalando.nakadi.client.scala.model.EventTypeCategory
import org.zalando.nakadi.client.scala.model.EventTypeSchema
import org.zalando.nakadi.client.scala.model.JacksonJsonMarshaller
import org.zalando.nakadi.client.scala.model.PartitionStrategy
import org.zalando.nakadi.client.scala.model.SchemaType
import org.zalando.nakadi.client.utils.ClientBuilder
import scala.concurrent.Await
import scala.concurrent.duration.DurationInt
import scala.collection.mutable.ListBuffer

object EventCreationExample extends App {

  // 1. Create the client

  import org.zalando.nakadi.client.utils.ClientBuilder
  import org.zalando.nakadi.client.scala.Client
  val client: Client = ClientFactory.getScalaClient()

  // 2. Create a simple Meeting Event class
  case class MeetingsEvent(date: String, topic: String) extends Event

  //We define the Json representation of our Meeting Event.
  val schema: String = """
    { 
      'properties': 
      { 
        'date': { 'type': 'string' }, 
        'topic': { 'type': 'string'}  
      } 
    }""".replaceAll("'", "\"")

  // 3. Create the EventType, 
  // We need to create an eventType(a topic), where listeners 
  // can subscribe to, to get our published events.
  val eventTypeSchema = new EventTypeSchema(SchemaType.JSON, schema)

  //First the eventType name, wich will be part of the URL: https://nakadi.test.io/event-types/{eventTypeName}
  //See the API for more information on the EventType model
  //https://github.com/zalando/nakadi/blob/nakadi-jvm/api/nakadi-event-bus-api.yaml#L1240
  //  val eventTypeName = "Example-unique-million-messages"
  val eventTypeName = "Example-2000"
//  val eventTypeName = "Example-unique-hundred-messages-3"

  val owner = "team-laas"
  val category = EventTypeCategory.UNDEFINED // We want just to pass data without through Nakadi, simple schema-validation is enough!
  val enrichmentStrategies = Nil
  val partitionStrategy = Some(PartitionStrategy.RANDOM)
  val dataKeyFields = Nil
  val paritionKeyFields = List("date", "topic")

  val eventType = new EventType(eventTypeName,
    owner,
    category,
    enrichmentStrategies,
    partitionStrategy,
    eventTypeSchema,
    dataKeyFields,
    paritionKeyFields,
    None)

  //You need to import the default Serializer if you don't sepecify your own!
  import JacksonJsonMarshaller._

  client.createEventType(eventType)
  Thread.sleep(3000)
  // 4. Publish the EventType
//  System.exit(0)
  var counter = 0
  for (n <- 1 to 1) {
    val event = new MeetingsEvent("2016-04-28T13:28:15+00:00", "Hackaton")
    var events = ListBuffer[MeetingsEvent]()
    for (a <- 1 to 1) {
      counter += 1
      events += MeetingsEvent("2016-04-28T13:28:15+00:00", "Hackaton" + counter)
    }
    Await.result(client.publishEvents(eventTypeName, events), 120.seconds)
  }
  client.stop()

}

File: it/src/main/java/org/zalando/nakadi/client/examples/java/EventCreationExample.java
package org.zalando.nakadi.client.examples.java;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

import org.zalando.nakadi.client.java.Client;
import org.zalando.nakadi.client.java.enumerator.EventEnrichmentStrategy;
import org.zalando.nakadi.client.java.enumerator.EventTypeCategory;
import org.zalando.nakadi.client.java.enumerator.PartitionStrategy;
import org.zalando.nakadi.client.java.enumerator.SchemaType;
import org.zalando.nakadi.client.java.model.Event;
import org.zalando.nakadi.client.java.model.EventType;
import org.zalando.nakadi.client.java.model.EventTypeSchema;
import org.zalando.nakadi.client.java.model.EventTypeStatistics;
import org.zalando.nakadi.client.java.utils.SerializationUtils;
import org.zalando.nakadi.client.scala.ClientFactory;

import com.google.common.collect.Lists;

public class EventCreationExample {
	/**
	 * Define how your event should look like
	 */
	public class MeetingsEvent implements Event {
		private final String date;
		private final String topic;

		public MeetingsEvent(String date, String topic) {
			this.date = date;
			this.topic = topic;
		}

		public String getDate() {
			return date;
		}

		public String getTopic() {
			return topic;
		}

	}

	public EventTypeSchema createEventTypeSchema(String schema) {
		return new EventTypeSchema(SchemaType.JSON, schema);
	}

	public EventType createEventType(String name,
			EventTypeSchema eventTypeSchema) {
		String owningApplication = "team-laas";
		EventTypeCategory category = EventTypeCategory.UNDEFINED;
		List<EventEnrichmentStrategy> enrichmentStrategies = Lists
				.newArrayList();
		PartitionStrategy partitionStrategy = PartitionStrategy.RANDOM;

		List<String> dataKeyFields = null;
		List<String> partitionKeyFields = Lists.newArrayList("date", "topic");
		EventTypeStatistics statistics = null;
		return new EventType(name, //
				owningApplication, //
				category, //
				enrichmentStrategies, //
				partitionStrategy, //
				eventTypeSchema, //
				dataKeyFields, //
				partitionKeyFields, //
				statistics);

	}

	public static void main(String[] args) throws InterruptedException,
			ExecutionException {

		/**
		 * An Event-type is ment for describing: 1. The Type of Events you want
		 * to create (i.e.: BusinessEvent) . 2. Schema validations to be
		 * enforced(or not)by Nakadi 3. How events should be distributed between
		 * their own partitions 4. A unique identifier for Subscribing
		 * 
		 */
		String eventTypeName = "Example-"+System.currentTimeMillis();

		/**
		 * Create client
		 */
		final Client client = ClientFactory.buildJavaClient();

		/**
		 * nakadi needs to know what kind of Json-schema you are going to send to
		 * the Event. We need to define a schema that matches the Event that we
		 * want to send along(MeetingsEvent).
		 */
		String schema = " { 'properties': { 'date': { 'type': 'string' }, 'topic': { 'type': 'string'} } }"
				.replaceAll("'", "\"");

		EventCreationExample example = new EventCreationExample();
		EventTypeSchema eventTypeSchema = example.createEventTypeSchema(schema);
		EventType eventType = example.createEventType(eventTypeName,
				eventTypeSchema);
		Future<Void> result = null;
		result = client.createEventType(eventType);
		result.get();

		// Create the event
		MeetingsEvent event = example.new MeetingsEvent(
				"2016-04-28T13:28:15+00:00", "Hackaton");
		MeetingsEvent event1 = example.new MeetingsEvent(
				"2016-04-28T13:28:15+00:00", "Hackaton1");
		MeetingsEvent event2 = example.new MeetingsEvent(
				"2016-04-28T13:28:15+00:00", "Hackaton2");
		MeetingsEvent event3 = example.new MeetingsEvent(
				"2016-04-28T13:28:15+00:00", "Hackaton3");
		// Single Event
		result = client.publishEvent(eventTypeName, event);
		result.get();
		// Single Event with Serializer,
		result = client.publishEvent(eventTypeName, event1,
		        SerializationUtils.defaultSerializer());
		result.get();
		// Multi Event
		result = client.publishEvents(eventTypeName, Arrays.asList(event2));
		result.get();
		// Multi Event with Serializer
		
		
		result = client.publishEvents(eventTypeName, Arrays.asList(event3), SerializationUtils.defaultSerializer());
		result.get();
		client.stop();

	}
}
